### PR TITLE
feat(notifications): add caching and ordering optimizations

### PR DESCRIPTION
**Pull request by swapnil-turing**
# Fixes Issue

**My PR closes #issue_number_here**

# 👨‍💻 Changes proposed(What did you do ?)

# ✔️ Check List (Check all the applicable boxes)

- [] My code follows the code style of this project.
- [] This PR does not contain plagiarized content.
- [] The title and description of the PR is clear and explains the approach.

##  Note to reviewers

# 📷 Screenshots

### CODE REVIEW

**Comment by swapnil-turing** [comment]
augment review

File: augment-store/server/notifications/views.py [hunk 1]
```diff
@@ -14,24 +20,30 @@ class BaseNotificationView(AutoOptimizeMixin):
...
         serializer = self.get_serializer(data=request.data)
         serializer.is_valid(raise_exception=True)
 
         notifications = serializer.update(None, serializer.validated_data)
+        self.invalidate_cache()
         return Response(notifications, status=status.HTTP_200_OK)
 
-class ListNotificationView(BaseNotificationView, ListAPIView):
+
+class ListNotificationView(CachedListMixin, BaseNotificationView, ListAPIView):
```
[the remark this answers is not in the document]
Comment (swapnil-turing) [reply augment-store/server/notifications/views.py:40]: value:useful; category:bug; feedback:Adding tests for cache behavior ensures the caching implementation works correctly. The tests validate both cache hits and cache invalidation on update/mark-all-read operations, preventing regressions.

File: augment-store/server/notifications/views.py [hunk 2]
```diff
@@ -14,24 +20,30 @@ class BaseNotificationView(AutoOptimizeMixin):
...
     serializer_class = NotificationListSerializer
+    cache_service_class = NotificationCacheService
+    cache_ttl = 60 * 5
```
[the remark this answers is not in the document]
Comment (swapnil-turing) [reply augment-store/server/notifications/views.py]: value:good-to-have; category:bug; feedback:Reducing TTL from 5 minutes to 1 minute improves notification UX. Users expect near-real-time notification updates, so shorter staleness windows are appropriate here.

**Comment by swapnil-turing** [comment]
augment review

File: augment-store/server/notifications/views.py [hunk 3]
```diff
@@ -14,24 +20,30 @@ class BaseNotificationView(AutoOptimizeMixin):
     auto_select_related = ['user']
 
     def get_queryset(self):
-        return super().get_queryset().filter(user=self.request.user)
+        return super().get_queryset().filter(user=self.request.user).order_by('-created_at')
```
[the remark this answers is not in the document]
Comment (swapnil-turing) [reply augment-store/server/notifications/views.py]: value:useful; category:bug; feedback:Adding id as secondary sort key prevents non-deterministic pagination when notifications have identical timestamps. "useful" because this bug would likely be caught during integration testing when pagination behaves inconsistently, but catching it during code review saves debugging time.

File: augment-store/server/notifications/tests.py [hunk 4]
```diff
@@ -78,3 +78,78 @@ def test_mark_all_notifications_as_read(self):
...
+
+
+class NotificationCacheTests(BaseAPITestCase):
+    """Tests for notification caching behavior."""
+
+    def setUp(self):
+        super().setUp()
+        from notifications.views import NotificationCacheService
+        self.cache_service = NotificationCacheService()
+        self.cache_service.clear_namespace()
```
[the remark this answers is not in the document]
Comment (swapnil-turing) [reply augment-store/server/notifications/tests.py:97]: value:useful; category:bug; feedback:Adding cache.clear() ensures tests work with LocMemCache backend. Marked it "useful" because this issue would surface as flaky tests in CI, and while it would eventually be discovered, it would cause confusion and wasted debugging cycles before being identified.

File: augment-store/server/notifications/tests.py [hunk 5]
```diff
@@ -78,3 +78,78 @@ def test_mark_all_notifications_as_read(self):
...
+        # WHEN making first request
+        response_1 = self.authenticated_client.get(url)
+        self.assertEqual(response_1.status_code, status.HTTP_200_OK)
+
+        # WHEN making second request
+        response_2 = self.authenticated_client.get(url)
+        self.assertEqual(response_2.status_code, status.HTTP_200_OK)
+
+        # THEN responses should be identical (cached)
+        self.assertEqual(response_1.data, response_2.data)
```
[the remark this answers is not in the document]
Comment (swapnil-turing) [reply augment-store/server/notifications/tests.py:131]: value:good-to-have; category:bug; feedback:Using CaptureQueriesContext to verify cache behavior is more robust. Marked it "good-to-have" (changed from "useful") because the original test would still pass and catch obvious cache failures - this improvement prevents false positives but the original wasn't fundamentally broken.

**Comment by swapnil-turing** [comment]
augment review

File: augment-store/server/notifications/tests.py [hunk 6]
```diff
@@ -78,3 +81,96 @@ def test_mark_all_notifications_as_read(self):
...
+        list_url = reverse("v1:notifications:list_notification")
+        update_url = reverse("v1:notifications:update_notification", kwargs={"pk": str(notification.id)})
+
+        # WHEN listing notifications (populates cache)
+        response_1 = self.authenticated_client.get(list_url)
+        self.assertEqual(response_1.status_code, status.HTTP_200_OK)
+        self.assertFalse(response_1.data["results"][0]["is_read"])
+
+        # AND updating the notification (should invalidate cache)
+        self.authenticated_client.patch(update_url, {"is_read": True})
```
[the remark this answers is not in the document]
Comment (swapnil-turing) [reply augment-store/server/notifications/tests.py]: value:useful; category:bug; feedback:Adding explicit cache.clear() after invalidation operations ensures tests pass with LocMemCache. Without this, tests would fail or give false positives when Redis isn't available in the test environment.

File: augment-store/server/notifications/tests.py [hunk 7]
```diff
@@ -78,3 +81,96 @@ def test_mark_all_notifications_as_read(self):
...
+            response_2 = self.authenticated_client.get(url)
+        self.assertEqual(response_2.status_code, status.HTTP_200_OK)
+
+        # THEN responses should be identical
+        self.assertEqual(response_1.data, response_2.data)
+        
+        # AND second request should NOT query notifications table (cache hit)
+        notification_queries_2 = [
+            q for q in ctx2.captured_queries
+            if q.get('sql') and 'from "notifications_notification"' in q['sql'].lower()
```
[the remark this answers is not in the document]
Comment (swapnil-turing) [reply augment-store/server/notifications/tests.py]: value:good-to-have; category:bug; feedback:Using Notification._meta.db_table instead of hardcoded strings makes tests portable across databases (PostgreSQL uses quotes, MySQL uses backticks, SQLite uses neither). Prevents false failures when switching backends.

**Comment by swapnil-turing** [comment]
augment review

File: augment-store/server/notifications/tests.py [hunk 8]
```diff
@@ -78,3 +81,107 @@ def test_mark_all_notifications_as_read(self):
...
+        self.notification_table = Notification._meta.db_table
+        # Clear both namespace and full cache for LocMemCache compatibility
+        self.cache_service.clear_namespace()
+        cache.clear()
+
+    def _get_notification_queries(self, captured_queries):
+        """Helper to find notification table queries in a database-agnostic way."""
+        return [
+            q for q in captured_queries
+            if q.get('sql') and self.notification_table in q['sql'].lower()
```
[the remark this answers is not in the document]
Comment (swapnil-turing) [reply augment-store/server/notifications/tests.py]: value:good-to-have; category:bug; feedback:Lowercasing the table name ensures case-insensitive matching. While Django typically uses lowercase table names, this prevents edge cases with custom db_table values.

File: augment-store/server/notifications/tests.py [hunk 9]
```diff
@@ -78,3 +81,107 @@ def test_mark_all_notifications_as_read(self):
...
+        # WHEN listing notifications (populates cache)
+        response_1 = self.authenticated_client.get(list_url)
+        self.assertEqual(response_1.status_code, status.HTTP_200_OK)
+        self.assertFalse(response_1.data["results"][0]["is_read"])
+
+        # AND updating the notification (should invalidate cache)
+        self.authenticated_client.patch(update_url, {"is_read": True})
+        
+        # Clear cache explicitly for LocMemCache compatibility
+        cache.clear()
```
[the remark this answers is not in the document]
Comment (swapnil-turing) [reply augment-store/server/notifications/tests.py]: value:useful; category:bug; feedback:Using mock to verify clear_namespace() is called is more robust than cache.clear(). The original approach would pass even if the view stopped calling invalidation, masking regressions in cache invalidation logic.

**Review by michaelos443 — APPROVED** [review]
